src/internal/pjs/server_internal_test.go

```diff
@@ -0,0 +1,480 @@
+			req.Program = programA
+		})
+		eg.Go(func() error {
+			_, err := runJobFrom(t, egCtx, c, createRespA.Id.Id, programA, func(resp *pjs.ProcessQueueResponse) error {
+				jobAContext = resp.Context
+				close(jobAProcessing)
+				time.Sleep(5 * time.Second)
+				return nil
+			})
+			return err
+		})
+		<-jobAProcessing
+
+		programB := createFileSet(t, fc, map[string][]byte{"program": []byte("B")})
+		createRespB := createJob(t, ctx, c, fc, func(req *pjs.CreateJobRequest) {
+			req.Context = jobAContext
+			req.Program = programB
+		})
+		var jobBContext string
+		eg.Go(func() error {
+			_, err := runJobFrom(t, egCtx, c, createRespB.Id.Id, programB, func(resp *pjs.ProcessQueueResponse) error {
+				jobBContext = resp.Context
+				close(jobBProcessing)
+				time.Sleep(5 * time.Second)
+				return nil
+			})
+			return err
+		})
+		<-jobBProcessing
+
+		programC := createFileSet(t, fc, map[string][]byte{"program": []byte("C")})
+		createRespC := createJob(t, ctx, c, fc, func(req *pjs.CreateJobRequest) {
+			req.Context = jobBContext
+			req.Program = programC
+		})
+		eg.Go(func() error {
+			_, err := runJobFrom(t, egCtx, c, createRespC.Id.Id, programC, func(resp *pjs.ProcessQueueResponse) error {
+				close(jobCProcessing)
+				time.Sleep(5 * time.Second)
+				return nil
+			})
+			return err
+		})
+		<-jobCProcessing
+
+		// Cancel job A while A, B, and C are all processing
+		_, err := c.CancelJob(ctx, &pjs.CancelJobRequest{
+			Job: &pjs.Job{
+				Id: createRespA.Id.Id,
+			},
+		})
+		require.NoError(t, err)
+
+		require.NoError(t, eg.Wait())
+
+		inspectCancelledJob(t, ctx, c, createRespA.Id.Id)
+		inspectCancelledJob(t, ctx, c, createRespB.Id.Id)
+		inspectCancelledJob(t, ctx, c, createRespC.Id.Id)
+	})
+}
+
+func TestWalkJob(t *testing.T) {
+	c, fc := setupTest(t)
+	ctx := pctx.TestContext(t)
+	depth := 3
+	fullBinaryJobTree(t, ctx, depth, c, fc)
+
+	expectedSize := int64(math.Pow(2, float64(depth)) - 1)
+	expected := make(map[int64][]int64)
+	for i := int64(1); i <= expectedSize; i++ {
+		left := 2 * i
+		if left <= expectedSize {
+			expected[i] = append(expected[i], left)
+		}
+		right := 2*i + 1
+		if right <= expectedSize {
+			expected[i] = append(expected[i], right)
+
+		}
+	}
+	walkC, err := c.WalkJob(ctx, &pjs.WalkJobRequest{
+		Job: &pjs.Job{
+			Id: 1,
+		},
+		Algorithm: pjs.WalkAlgorithm_LEVEL_ORDER,
+	})
+	require.NoError(t, err)
+	actual := make(map[int64][]int64)
+	for {
+		resp, err := walkC.Recv()
+		if err != nil {
+			if errors.As(err, io.EOF) {
+				break
+			}
+			require.NoError(t, err)
+		}
+		parentJob := resp.Details.JobInfo.ParentJob
+		if parentJob != nil && parentJob.Id != 0 {
+			actual[parentJob.Id] = append(actual[parentJob.Id], resp.Id.Id)
+		}
+		require.NoError(t, err)
+	}
+	require.NoDiff(t, expected, actual, nil)
+}
+
+// runJob does work through PJS.
+func runJob(t *testing.T, ctx context.Context, c pjs.APIClient, in *pjs.CreateJobRequest,
```

Comment:
This and other test helpers should start with `t.Helper()`.  This way, failures are reported from the perspective of the caller, not at the line in this function that failed.

Right now, the errors will be unhelpful either way, as you can't disambiguate random RPC errors from CreateJob or runJobFrom's various require.NoError calls.  Adding a description to each require.NoError() call would help with that.